File: phase0/candidates/oscar64/c64_timer.cpp
#include "c64_timer.hpp"

static volatile unsigned char * const CIA1_TIMER_A_LOW =
    (volatile unsigned char *)0xdc04;
static volatile unsigned char * const CIA1_TIMER_A_HIGH =
    (volatile unsigned char *)0xdc05;
static volatile unsigned char * const CIA1_TIMER_B_LOW =
    (volatile unsigned char *)0xdc06;
static volatile unsigned char * const CIA1_TIMER_B_HIGH =
    (volatile unsigned char *)0xdc07;
static volatile unsigned char * const CIA1_INTERRUPT_CONTROL =
    (volatile unsigned char *)0xdc0d;
static volatile unsigned char * const CIA1_CONTROL_A =
    (volatile unsigned char *)0xdc0e;
static volatile unsigned char * const CIA1_CONTROL_B =
    (volatile unsigned char *)0xdc0f;
static volatile unsigned char * const CIA2_INTERRUPT_CONTROL =
    (volatile unsigned char *)0xdd0d;
static volatile unsigned char * const VIC_CONTROL_1 =
    (volatile unsigned char *)0xd011;
static volatile unsigned char * const VIC_RASTER =
    (volatile unsigned char *)0xd012;
static volatile unsigned char * const VIC_SPRITE_ENABLE =
    (volatile unsigned char *)0xd015;

static const unsigned char CONTROL_FORCE_LOAD = 0x10;
static const unsigned char CONTROL_START = 0x01;
static const unsigned char TIMER_B_COUNTS_TIMER_A = 0x40;

static void wait_for_frame_start(void) {
    while (*VIC_RASTER == 0 && (*VIC_CONTROL_1 & 0x80) == 0) {
    }
    while (*VIC_RASTER != 0 || (*VIC_CONTROL_1 & 0x80) != 0) {
    }
}

void prepare_cia_timing(void) {
    *CIA1_INTERRUPT_CONTROL = 0x7f;
    (void)*CIA1_INTERRUPT_CONTROL;
    *CIA2_INTERRUPT_CONTROL = 0x7f;
    (void)*CIA2_INTERRUPT_CONTROL;
    *VIC_CONTROL_1 &= 0xef;
    *VIC_SPRITE_ENABLE = 0x00;
}

void start_cia_timer(void) {
    wait_for_frame_start();
    *CIA1_CONTROL_A = 0x00;
    *CIA1_CONTROL_B = TIMER_B_COUNTS_TIMER_A;

    *CIA1_TIMER_A_LOW = 0xff;
    *CIA1_TIMER_A_HIGH = 0xff;
    *CIA1_TIMER_B_LOW = 0xff;
    *CIA1_TIMER_B_HIGH = 0xff;

    *CIA1_CONTROL_A = CONTROL_FORCE_LOAD;
    *CIA1_CONTROL_B = TIMER_B_COUNTS_TIMER_A | CONTROL_FORCE_LOAD;
    *CIA1_CONTROL_B = TIMER_B_COUNTS_TIMER_A | CONTROL_START;
    *CIA1_CONTROL_A = CONTROL_START;
}

unsigned long stop_cia_timer(void) {
    *CIA1_CONTROL_A = 0x00;
    *CIA1_CONTROL_B = TIMER_B_COUNTS_TIMER_A;

    unsigned long remaining = *CIA1_TIMER_A_LOW;
    remaining |= (unsigned long)*CIA1_TIMER_A_HIGH << 8;
    remaining |= (unsigned long)*CIA1_TIMER_B_LOW << 16;
    remaining |= (unsigned long)*CIA1_TIMER_B_HIGH << 24;
    return 0xffffffffUL - remaining;
}

unsigned long measure_cia_boundary_overhead(void) {
    start_cia_timer();
    return stop_cia_timer();
}

File: phase0/candidates/rust/src/c64_timer.rs
const CIA1_TIMER_A_LOW: *mut u8 = 0xdc04 as *mut u8;
const CIA1_TIMER_A_HIGH: *mut u8 = 0xdc05 as *mut u8;
const CIA1_TIMER_B_LOW: *mut u8 = 0xdc06 as *mut u8;
const CIA1_TIMER_B_HIGH: *mut u8 = 0xdc07 as *mut u8;
const CIA1_INTERRUPT_CONTROL: *mut u8 = 0xdc0d as *mut u8;
const CIA1_CONTROL_A: *mut u8 = 0xdc0e as *mut u8;
const CIA1_CONTROL_B: *mut u8 = 0xdc0f as *mut u8;
const CIA2_INTERRUPT_CONTROL: *mut u8 = 0xdd0d as *mut u8;
const VIC_CONTROL_1: *mut u8 = 0xd011 as *mut u8;
const VIC_RASTER: *mut u8 = 0xd012 as *mut u8;
const VIC_SPRITE_ENABLE: *mut u8 = 0xd015 as *mut u8;

const CONTROL_FORCE_LOAD: u8 = 0x10;
const CONTROL_START: u8 = 0x01;
const TIMER_B_COUNTS_TIMER_A: u8 = 0x40;

unsafe fn write_register(address: *mut u8, value: u8) {
    core::ptr::write_volatile(address, value);
}

unsafe fn read_register(address: *mut u8) -> u8 {
    core::ptr::read_volatile(address)
}

unsafe fn wait_for_frame_start() {
    while read_register(VIC_RASTER) == 0 && read_register(VIC_CONTROL_1) & 0x80 == 0 {}
    while read_register(VIC_RASTER) != 0 || read_register(VIC_CONTROL_1) & 0x80 != 0 {}
}

pub unsafe fn prepare_cia_timing() {
    write_register(CIA1_INTERRUPT_CONTROL, 0x7f);
    let _ = read_register(CIA1_INTERRUPT_CONTROL);
    write_register(CIA2_INTERRUPT_CONTROL, 0x7f);
    let _ = read_register(CIA2_INTERRUPT_CONTROL);
    write_register(VIC_CONTROL_1, read_register(VIC_CONTROL_1) & 0xef);
    write_register(VIC_SPRITE_ENABLE, 0x00);
}

pub unsafe fn start_cia_timer() {
    wait_for_frame_start();
    write_register(CIA1_CONTROL_A, 0x00);
    write_register(CIA1_CONTROL_B, TIMER_B_COUNTS_TIMER_A);

    write_register(CIA1_TIMER_A_LOW, 0xff);
    write_register(CIA1_TIMER_A_HIGH, 0xff);
    write_register(CIA1_TIMER_B_LOW, 0xff);
    write_register(CIA1_TIMER_B_HIGH, 0xff);

    write_register(CIA1_CONTROL_A, CONTROL_FORCE_LOAD);
    write_register(CIA1_CONTROL_B, TIMER_B_COUNTS_TIMER_A | CONTROL_FORCE_LOAD);
    write_register(CIA1_CONTROL_B, TIMER_B_COUNTS_TIMER_A | CONTROL_START);
    write_register(CIA1_CONTROL_A, CONTROL_START);
}

pub unsafe fn stop_cia_timer() -> u32 {
    write_register(CIA1_CONTROL_A, 0x00);
    write_register(CIA1_CONTROL_B, TIMER_B_COUNTS_TIMER_A);

    let timer_a_low = read_register(CIA1_TIMER_A_LOW) as u32;
    let timer_a_high = read_register(CIA1_TIMER_A_HIGH) as u32;
    let timer_b_low = read_register(CIA1_TIMER_B_LOW) as u32;
    let timer_b_high = read_register(CIA1_TIMER_B_HIGH) as u32;
    let remaining = timer_a_low | (timer_a_high << 8) | (timer_b_low << 16) | (timer_b_high << 24);
    u32::MAX.wrapping_sub(remaining)
}

pub unsafe fn measure_cia_boundary_overhead() -> u32 {
    start_cia_timer();
    stop_cia_timer()
}
